File: 刷题/cpp/86. 分隔链表.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* partition(ListNode* head, int x) {
        ListNode* head1=new ListNode(-1);
        ListNode* head2=new ListNode(-1);
        ListNode *less=head1, *great=head2;
        ListNode* p=head;
        while(p!=nullptr){
            if(p->val<x){
                less->next=p;
                less=p;
            } else{
                great->next=p;
                great=p;
            }
            p=p->next;
        }
        great->next=nullptr;
        less->next=head2->next;
        return head1->next;
    }
};

File: 刷题/golang/86. 分隔链表.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
 func partition(head *ListNode, x int) *ListNode {
    dummy1:=&ListNode{-1,nil}
    dummy2:=&ListNode{-1,nil}
    small, large:=dummy1, dummy2
    p:=head
    for p!=nil {
        if p.Val < x {
            small.Next = p
            small = small.Next
        } else {
            large.Next = p
            large = large.Next
        }
        p=p.Next
    }
    large.Next=nil
    small.Next=dummy2.Next
    return dummy1.Next
}